controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1190 @@
+			}
+		}
+	}
+
+	for id, j := range s.jobs {
+		if _, ok := newJobs[id]; !ok && j.IsRunning() {
+			s.handleJobStop(j)
+		}
+	}
+
+	s.triggerRectify()
+}
+
+func (s *Scheduler) SyncFormations() {
+	defer s.sendEvent(NewEvent(EventTypeFormationSync, nil, nil))
+
+	log := fnLogger()
+	log.Info("syncing formations")
+
+	log.Info("getting app list")
+	apps, err := s.AppList()
+	if err != nil {
+		log.Error("error getting apps", "err", err)
+		return
+	}
+
+	for _, app := range apps {
+		appLog := log.New("app.id", app.ID)
+
+		fs, err := s.FormationList(app.ID)
+		if err != nil {
+			appLog.Error("error getting formations", "err", err)
+			continue
+		}
+		appLog.Info(fmt.Sprintf("got %d formation(s) for %s app", len(fs), app.Name))
+
+		for _, f := range fs {
+			if _, err := s.updateFormation(f); err != nil {
+				appLog.Error("error updating formation", "release.id", f.ReleaseID, "err", err)
+			}
+		}
+	}
+
+	s.triggerRectify()
+}
+
+func (s *Scheduler) Rectify() {
+	log := fnLogger()
+
+	if !s.isLeader {
+		log.Warn("ignoring rectify as not service leader")
+		return
+	}
+
+	defer s.sendEvent(NewEvent(EventTypeRectify, nil, nil))
+
+	log.Info("starting rectify")
+
+	for key, formation := range s.formations {
+		formationLog := log.New("app.id", key.AppID, "release.id", key.ReleaseID)
+		formationLog.Info("rectifying formation")
+
+		expected := formation.GetProcesses()
+		actual := s.jobs.GetProcesses(key)
+
+		if expected.Equals(actual) {
+			formationLog.Info("formation in correct state", "expected", expected, "actual", actual)
+			continue
+		}
+		// TODO: make this formation.Diff(actual)?
+		formation.Processes = actual
+		diff := formation.Update(expected)
+		formationLog.Info("existing formation in incorrect state", "expected", expected, "actual", actual, "diff", diff)
+		s.sendDiffRequests(formation, diff)
+	}
+}
+
+func (s *Scheduler) HandleFormationChange(ef *ct.ExpandedFormation) {
+	var err error
+	defer func() {
+		s.sendEvent(NewEvent(EventTypeFormationChange, err, nil))
+	}()
+
+	log := fnLogger()
+	if ef == nil || ef.App == nil || ef.Release == nil {
+		log.Error("invalid formation")
+		return
+	}
+
+	log = log.New("app.id", ef.App.ID, "release.id", ef.Release.ID)
+	log.Info("handling formation change")
+	_, err = s.changeFormation(ef)
+	if err != nil {
+		log.Error("error handling formation change", "err", err)
+		return
+	}
+
+	s.triggerRectify()
+}
+
+func (s *Scheduler) HandleJobRequest(req *JobRequest) {
+	log := fnLogger("req.id", req.JobID, "req.type", req.RequestType)
+
+	if !s.isLeader {
+		log.Warn("ignoring job request as not service leader")
+		return
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			log.Error("error handling job request", "err", err)
+		}
+		s.sendEvent(NewEvent(EventTypeJobRequest, err, req))
+	}()
+
+	log.Info("handling job request")
+	switch req.RequestType {
+	case JobRequestTypeUp:
+		// startJob sets the HostID on the request if successful
+		// otherwise it remains blank
```

Comment:
This comment is stale, from when I was manually handling state in `s.pendingStarts`